crml/nft/src/lib.rs

```diff
@@ -0,0 +1,651 @@
+		}
+
+		/// Place a bid on an open auction
+		/// - `amount` to bid (in the seller's requested payment asset)
+		#[weight = T::WeightInfo::bid()]
+		#[transactional]
+		fn bid(origin, collection_id: CollectionId, token_id: T::TokenId, amount: Balance) {
+			let origin = ensure_signed(origin)?;
+			ensure!(<Listings<T>>::contains_key(&collection_id, token_id), Error::<T>::NotForAuction);
+
+			if let Some(Listing::Auction(listing)) = Self::listings(&collection_id, token_id) {
+				if let Some(current_bid) = Self::listing_winning_bid(&collection_id, token_id) {
+					ensure!(amount > current_bid.1, Error::<T>::BidTooLow);
+				} else {
+					// first bid
+					ensure!(amount >= listing.reserve_price, Error::<T>::BidTooLow);
+				}
+
+				// check user has the requisite funds to make this bid
+				let balance = T::MultiCurrency::free_balance(&origin, Some(listing.payment_asset));
+				if let Some(balance_after_bid) = balance.checked_sub(amount) {
+					// TODO: review this during 3.0 upgrade
+					// - `amount` is unused
+					// - if there are multiple locks on user asset this could return true inaccurately
+					// - `T::MultiCurrency::reserve(origin, asset_id, amount)` should be checking this internally...
+					let _ = T::MultiCurrency::ensure_can_withdraw(&origin, Some(listing.payment_asset), amount, WithdrawReason::Reserve.into(), balance_after_bid)?;
+				}
+
+				// try lock funds
+				T::MultiCurrency::reserve(&origin, Some(listing.payment_asset), amount)?;
+
+				ListingWinningBid::<T>::mutate(&collection_id, token_id, |maybe_current_bid| {
+					if let Some(current_bid) = maybe_current_bid {
+						// replace old bid
+						T::MultiCurrency::unreserve(&current_bid.0, Some(listing.payment_asset), current_bid.1);
+					}
+					*maybe_current_bid = Some((origin, amount))
+				});
+
+				Self::deposit_event(RawEvent::Bid(collection_id, token_id, amount));
+			} else {
+				return Err(Error::<T>::NotForAuction.into());
+			}
+		}
+
+		/// Close a sale or auction
+		/// Requires no successful bids have been made for the auction.
+		/// Caller must be the token owner
+		#[weight = T::WeightInfo::cancel_sale()]
+		fn cancel_sale(origin, collection_id: CollectionId, token_id: T::TokenId) {
+			let origin = ensure_signed(origin)?;
+			let current_owner = Self::token_owner(&collection_id, token_id);
+			ensure!(current_owner == origin, Error::<T>::NoPermission);
+
+			match Self::listings(&collection_id, token_id) {
+				Some(Listing::<T>::DirectSale(sale)) => {
+					ListingEndSchedule::<T>::mutate(sale.close, |schedule| schedule.retain(|(c, t)| (c, t) != (&collection_id, &token_id)));
+					Listings::<T>::remove(&collection_id, token_id);
+					Self::deposit_event(RawEvent::DirectSaleClosed(collection_id, token_id));
+				},
+				Some(Listing::<T>::Auction(auction)) => {
+					ensure!(Self::listing_winning_bid(&collection_id, token_id).is_none(), Error::<T>::TokenListingProtection);
```

Comment:
If we decide to keep bids below the reserve price until the settlement, which I highly suggest, here we should check if there is a bid that meets the reserved price.